refactor errors

tjhance opened the pull request:
To support a more structured way of representing errors, this adds the `Error` type to the air AST.

```rust
pub struct ErrorX {
    pub msg: String,
    pub spans: Vec<Span>, // "primary" spans
    pub labels: Vec<Label>,
}
```

This allows us to easily format any error that can be displayed by converting to a rustc `MultiSpan` and printing with `span_diagnostic.span_err` (see `report_vir_error`). And I expect it will be easily extensible if need be.

Look at the SST -> AIR conversion diffs to see what it now looks like to create an `Error` object now. In the future we can change `VirErr` to use the same `Error` type as well. (EDIT: now done)

The main challenge in this diff is to integrate the AST change with the existing scheme in `smt_verify.rs` for discovering spans to add as additional labels. After discussion with Chris, it seemed to make sense to have two different node types in the air AST, one for a base error, and one for additional labels. Originally, I planned to just change `Assert` and `LabeledAssertion`. However, what I realized during the refactor is that the weakest-precondition code converts `Assert` statements to `LabelAssertions`, so they need to take the same type of extra info. I ended up doing this instead:

```rust
pub enum ExprX {
    ...
    LabeledAxiom(Labels, Expr),
    LabeledAssertion(Error, Expr),
}

pub enum StmtX {
    ...
    Assert(Error, Expr),
}
```

As you can see, I split the previous `LabelledAssertion` into two. I'm not really sure about the naming ... I called it `LabeledAxiom` because it seemed like the labels were only relevant to axioms. Open to suggestions here.

In the PR, my changes to `label_asserts` are the most in need of review; since I split the node into two types, I just made it the case that it only looks for `LabeledAxioms` in `global` mode and `LabeledAssertions` otherwise. This seemed to make sense based on my read of the code, but I wasn't sure.

utaal commented on source/air/src/ast.rs:
```diff
@@ -4,11 +4,30 @@ use std::sync::Arc;
+#[derive(Clone)] // for Debug, see ast_util
+pub struct ErrorX {
+    pub msg: String,
+    pub spans: Vec<Span>, // "primary" spans
```
My understanding is that "primary" is the code that caused the error, non-primary spans are the labels that refer to other locations used in the error message (like in the example above). I like this (and I was thinking I'd like something like this for error messages I'll be writing). These multi-span errors are common in rust for borrowck failures: there's a span that's misusing a variable (e.g. transferring ownership of something that's currently borrowed), the "primary" here, and two "label" spans that point to the point where the borrow starts and ends, which is very useful.

tjhance replied:
I added an example of an error msg and what each field corresponds to

parno replied:
Thanks, the example and the extra documentation is quite helpful!  Can you add a comment for when you might have multiple primary spans?

tjhance commented on source/air/src/ast.rs:
```diff
@@ -4,11 +4,30 @@ use std::sync::Arc;
+/// with an additional 'msg' String to serve as an error message.
+/// An Error should always have at least one 'span' which represents
+/// the primary point where the error is. It is possible to have more
+/// than one span, and it is possible to have additional label information.
```
a Label is a (span, string) pair. It's independent of the 'spans' field

parno replied:
Got it, thanks.

Chris-Hawblitzel replied:
Maybe we should call the struct `ErrorLabel` for clarity?

utaal commented:
This looks great to me. Are you planning these so we can wire them up to be able to use multiple spans when emitting errors at the Rust->VIR stage?

tjhance commented:
yes, I just added a commit for that

Chris-Hawblitzel commented on source/air/src/parser.rs:
```diff
@@ -94,9 +94,9 @@ impl Parser {
+                        let spans = self.nodes_to_labels(labels)?;
                         let expr = self.node_to_expr(e)?;
-                        return Ok(Arc::new(ExprX::LabeledAssertion(spans, expr)));
+                        return Ok(Arc::new(ExprX::LabeledAxiom(spans, expr)));
```
The printer has cases for both `LabeledAxiom` and `LabeledAssertion`, but I only see `LabeledAxiom` in the parser.  We should be able to round-trip between the printer and parser.  Should we have separate syntax, like `axiom_location` and `location` to distinguish these?

Chris-Hawblitzel commented on source/air/src/smt_verify.rs:
```diff
@@ -11,6 +11,7 @@ use std::sync::Arc;
     infos: &mut Vec<AssertionInfo>,
+    axiom_infos: &mut Vec<AxiomInfo>,
     expr: &Expr,
     is_global: bool,
```
Can `is_global` be removed here?

tjhance commented:
@Chris-Hawblitzel I've addressed your suggestions.

Regarding `is_global`, I expect you know better than I do. It kinda sounds like in `global=true` mode it should only find LabeledAxioms and not any LabeledAssertions, and vice versa? So the question is what should happen if that weren't true, should it process them as normal or skip them?

Chris-Hawblitzel commented:
It's fine to just process them.  It won't do anything useful as far as I know, but it won't cause any harm.
